### ISSUE TITLE
is it possible to control edge color?

### ISSUE THREAD

**Issue by JosephBARBIERDARNAL** [gyli/PyWaffle#37]
Hello!

Is there a way to control the border of each block? Because if there isn't, I get the impression that it's not possible to create dark, satisfying graphics with pywaffle.

Here is an example (and its output):

```python
import matplotlib.pyplot as plt
import pandas as pd
from pywaffle import Waffle

path = 'https://raw.githubusercontent.com/holtzy/R-graph-gallery/master/DATA/share-cereals.csv'
df = pd.read_csv(path)

def remove_html_tag(s):
    return s.split('</b>')[0][3:]

df['lab'] = df['lab'].apply(remove_html_tag)
df = df[df['type'] == 'feed']

background_color = "#222725"
pink = "#f72585"
black = "#4F0325"

number_of_bars = len(df)  # one bar per continent

# Init the whole figure and axes
fig, axs = plt.subplots(
   nrows=number_of_bars,
   ncols=1,
   figsize=(8, 6)
)
fig.set_facecolor(background_color)
ax.set_facecolor(background_color)


# Iterate over each bar and create it
for (i, row), ax in zip(df.iterrows(), axs):

    share = row['percent']
    values = [share, 100-share]

    Waffle.make_waffle(
        ax=ax,
        rows=4,
        columns=25,
        values=values,
        colors=[pink, black],
        edgecolor='white'
    )

plt.show()
```

<img width="763" alt="Screenshot 2024-07-16 at 14 25 50" src="https://github.com/user-attachments/assets/193a1d4c-9351-48be-953b-ffc2f392fa0f">

**Comment by JosephBARBIERDARNAL** [comment]
I'd like to be able to make something like:

<img width="479" alt="Screenshot 2024-07-16 at 14 29 46" src="https://github.com/user-attachments/assets/1ff1f96b-f56f-4959-95f6-d61f2aed2fec">

**Comment by gyli** [comment]
Hi @JosephBARBIERDARNAL! I think that's a nice to have feature, let me see if and how it can be aded.

**Comment by JosephBARBIERDARNAL** [comment]
cool! I'm ready to help if you need it

**Comment by gyli** [comment]
PR is surely welcomed!
While before starting coding, I think it's discussable how to achieve this. 
At my first glance, I thought we just need to add 4 edges outside of each block. But what would be the default line weight? Seems like the screenshot you attached is a special case, that edges happen to fill the gaps, and I think we can make this as the default option. 
Then, how can we make the edge weight adjustable? So I can have block with edge and gap at the same time? Probably we can make a new parameter like `edge_ratio` (the ratio of edge width and gap width), similar to how to control gap size with `interval_ratio_x` and `interval_ratio_y`. 
Also, [the block might not always be a square](https://pywaffle.readthedocs.io/en/latest/examples/block_shape_spacing_location_direction_and_style.html#block-spacing), and I don't have a good solution to add beautiful edges in this case. Probably we can skip this case.

**Comment by JosephBARBIERDARNAL** [comment]
Hello again!

My first intuition is as follows
- let's start by having a simple way of filling in the edge colors (similar to my example above)
- the default colour would be the background color of the axes (which is currently the case)?
- This means that regardless of the shape of the blocks or the space between them, all the empty spaces between the blocks will be filled

then, depending on the usefulness you see in it, a way of controlling the weight of the edges (which means leaving partially empty spaces?)

but IMO the main use case would be to fill the spaces using a given color. Since I haven't looked at the source code yet, I don't measure how complex is it to implement?

_Note: I'm don't have much free time at the time I'm writing this, but I will have much more from September (:_

**Comment by gyli** [comment]
@JosephBARBIERDARNAL — sorry for the long silence here. Coming back to this with a decision rather than more design questions.

**I think your instinct was the better one.** I proposed `edge_ratio` earlier, coupling edge width to the gap width; you suggested just filling the space behind the blocks. Yours is simpler, works for any `block_aspect_ratio` and any `interval_ratio_x/y`, and needs no per-block geometry — which was exactly the case I said I had no good solution for.

So rather than one coupled parameter, two independent ones:

- **`background_color`** — a single `Rectangle` spanning the axes limits, drawn first at low `zorder`. This is your screenshot: the gaps between blocks get a colour. Works with any block shape or spacing, and it is the main use case you described.
- **`block_edge_color`** / **`block_edge_width`** — passed straight through to the `Rectangle` as `edgecolor` and `linewidth`. Nearly free, since the `Rectangle` is already being constructed and matplotlib does the drawing.

Keeping them separate means you can have a background fill *and* visible gaps, or edges without a fill, rather than one ratio deciding both.

One limitation worth stating up front: `block_edge_color` applies only to rectangle blocks. The `icons` and `characters` modes draw `Text` artists, which have no edge in this sense. `background_color` works for all three.

**A PR is very welcome** if you still have the time and interest — and if not, say so and I will pick it up; either way you should not be waiting on me again. If you do take it, the place to look is the block-drawing loop in `pywaffle/waffle.py` and `_default_parameters` just above it.

Two things that have changed since you filed this, both worth pulling before you start:

- #38 is open and touches the same function, including moving all argument validation into `_parameter_validation`. New parameters should be validated there.
- The test suite went from 5 tests to 56 with block-geometry assertions, so there is now something real to write a test against.

Thanks for the original suggestion and for the patience.

**Comment by gyli** [comment]
This will be rollout in the next version, 1.2.0
